Python/compile.c

```diff
@@ -6493,6 +6493,29 @@ optimize_cfg(struct assembler *a, PyObject *consts)
             b->b_iused = 0;
        }
     }
+    /* Delete jump instructions made redundant by previous step. If a non-empty
+       block ends with a jump instruction, check if the next non-empty block
+       reached through normal flow control is the target of that jump. If it
+       is, then the jump instruction is redundant and can be deleted.
+    */
+    for (basicblock *b = a->a_entry; b != NULL; b = b->b_next) {
+        if (b->b_iused > 0) {
+            struct instr *b_last_instr = &b->b_instr[b->b_iused - 1];
+            if (b_last_instr->i_opcode == JUMP_ABSOLUTE ||
+                b_last_instr->i_opcode == JUMP_FORWARD) {
+                basicblock *b_next_act = b->b_next;
+                while (b_next_act != NULL && b_next_act->b_iused == 0) {
+                    /* Next line reqd to prevent assertion failure in stackdepth() */
```

Comment:
I'm not sure that this is correct. What if the empty blocks are reachable by some other path? Changing `b_nofallthrough` on them could be incorrect.

An alternative approach would be to update all `b->b_next` and jump targets on all blocks to skip the empty blocks.
Then the CFG would remain correct without having to modify the  intermediate blocks, which may or may not be reachable.

Reply:
Maybe I'm not understanding the purpose of `b_nofallthrough` correctly, but even if those empty blocks are reachable by some other path, shouldn't `b_nofallthrough` still be set to 0 for them? I mean, if they're empty, then by definition they can't have jumps that lead anywhere else, right? So they should only have one exit path and that too only to the next block. If we're traversing them, there's no path to take other than falling through to the next block if they're empty.

Actually this made me realize that `b_reachable` should also be set back to 1 for these blocks once the jump is gone, since whether they're reachable by some other path or not, they're definitely reachable now from block in which the jump was deleted. Technically `b_nofallthrough` should also only modified if the jump is to be deleted, but again unless I'm grossly misunderstanding something having it as 0 should be the correct state for empty blocks anyway, so it shouldn't matter.

Reply:
If a block is unreachable, `stackdepth()` shouldn't see it, so if there is an assertion failure in stackdepth(), it implies that the block is reachable.
If a block is reachable *and* empty, the `b_nofallthrough` should be 0 already.

I think the problem might be that we assume all CFG modifications are done when we delete all the unreachable code.
So the resulting CFG might be invalid (in a way that doesn't matter to the assembler), once all unreachable code is deleted. In order to maintain correctness of the CFG, `b->b_nofallthrough = 0` should probably go [here](https://github.com/python/cpython/pull/23733/files#diff-ebc983d9f91e5bcf73500e377ac65e85863c4f77fd5b6b6caf4fcdf7c0f0b057R6493)

Reply:
That makes sense, thanks! I'll move the `b->b_nofallthrough = 0` statement and correct `b_reachable` on the affected blocks.